brave/src/main/java/brave/RealSpan.java

```diff
@@ -0,0 +1,219 @@
+package brave;
+
+import com.google.auto.value.AutoValue;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+import zipkin.Constants;
+import zipkin.Endpoint;
+import zipkin.reporter.Reporter;
+
+/**
+ * Offset-based span: Uses a single point of reference and offsets to create annotation timestamps.
+ *
+ * <p>Internally, state is kept in a way portable with Zipkin 1 and Zipkin 2 representations.
+ */
+final class RealSpan extends Span {
+  static final Map<String, String> END_ANNOTATION_TO_ADDRESS_TYPE;
+
+  static {
+    Map<String, String> result = new LinkedHashMap<>();
+    result.put(Constants.CLIENT_RECV, Constants.SERVER_ADDR);
+    result.put(Constants.SERVER_SEND, Constants.CLIENT_ADDR);
+    // TODO: add mr/ms -> ma when rfc is finalized
+    END_ANNOTATION_TO_ADDRESS_TYPE = Collections.unmodifiableMap(result);
+  }
+
+  final ZipkinRecorder recorder;
+  final Clock clock;
+  final Ticker ticker;
+  final TraceContext context;
+
+  // all ticks are relative to these
+  final long createTimestamp;
+  final long createTick;
+
+  // initialize to max so it is easy to compare
+  final AtomicLong timestamp = new AtomicLong(Long.MAX_VALUE);
+
+  // this lock guards all the below fields.
+  final Object lock = new Object();
+
+  String name;
+  /** even indexing timestamp:value */
+  List<Object> annotations;
+  Map<String, String> tags;
+  String addressType;
+  Endpoint remoteEndpoint;
+  volatile boolean finished;
+
+  RealSpan(Factory factory, TraceContext context) {
+    this.recorder = new ZipkinRecorder(factory.localEndpoint(), factory.reporter());
+    this.clock = factory.clock();
+    this.ticker = factory.ticker();
+    this.context = context;
+    this.createTimestamp = clock.epochMicros();
+    this.createTick = ticker.tickNanos();
+  }
+
+  @Override public boolean isNoop() {
+    return false;
+  }
+
+  @Override public TraceContext context() {
+    return context;
+  }
+
+  @Override public Span start() {
+    setIfLessThanCurrentTimestamp(relativeTimestamp());
+    return this;
+  }
+
+  @Override public Span start(long timestamp) {
+    setIfLessThanCurrentTimestamp(timestamp);
+    return this;
+  }
+
+  @Override public Span setName(String name) {
+    if (name == null) throw new NullPointerException("name == null");
+    synchronized (lock) {
+      this.name = name;
+    }
+    return this;
+  }
+
+  @Override public Span addAnnotation(String value) {
+    return addAnnotation(relativeTimestamp(), value);
+  }
+
+  @Override public Span addAnnotation(long timestamp, String value) {
+    if (value == null) throw new NullPointerException("value == null");
+    setIfLessThanCurrentTimestamp(timestamp);
+    synchronized (lock) {
+      if (annotations == null) annotations = new ArrayList<>();
+      annotations.add(timestamp);
+      annotations.add(value);
```

Comment:
Why are these not tied together by some object? It would seem more natural and less error prone if `annotations` was a `List<Pair>` or something along those lines.